src/context/node.js

```diff
@@ -0,0 +1,354 @@
+ * are auto-discovered or prompted.
+ *
+ * @package
+ */
+export class ContextNode {
+  /**
+   * Returns the existing context node or creates a new one.
+   *
+   * @param {!Node} node
+   * @return {!ContextNode}
+   */
+  static get(node) {
+    let contextNode = /** @type {!ContextNode|undefined} */ (node[NODE_PROP]);
+    if (!contextNode) {
+      contextNode = new ContextNode(node);
+      if (getMode().localDev || getMode().test) {
+        // The `Object.defineProperty({enumerable: false})` helps tests, but
+        // hurts performance. So this is only done in a dev/test modes.
+        Object.defineProperty(node, NODE_PROP, {
+          value: contextNode,
+          writable: false,
+          enumerable: false,
+          configurable: false,
+        });
+      } else {
+        node[NODE_PROP] = contextNode;
+      }
+    }
+    return contextNode;
+  }
+
+  /**
+   * Returns the closest available context node to the one specified. If the
+   * `node` has the context node, it's returned unless `includeSelf` is set as
+   * `false`.
+   *
+   * The DOM traversal goes at most as far as the root node (document,
+   * shadow root, document fragment) or as far as the DOM tree allows. The
+   * traversal follows the assigned shadow slots.
+   *
+   * Root nodes (document or shadow root) and AMP elements are auto-created
+   * during the traversal.
+   *
+   * @param {!Node} node The node from which to perform the search.
+   * @param {boolean=} includeSelf Whether the specified node itself should
+   * be included in the search. Defaults to `true`.
+   * @return {?ContextNode}
+   */
+  static closest(node, includeSelf = true) {
+    let n = node;
+    while (n) {
+      // Check if a node is a candidate to be returned.
+      if (n != node || includeSelf) {
+        if (n[NODE_PROP]) {
+          // Already a discovered node.
+          return /** @type {!ContextNode} */ (n[NODE_PROP]);
+        }
+        const {nodeType} = n;
+        if (nodeType == DOCUMENT_NODE || nodeType == FRAGMENT_NODE) {
+          // A context node is always created for a root. Due to this, a
+          // non-root element is always at least attached to a root. This
+          // allows for quick discovery and reattachment when new information
+          // becomes available.
+          return ContextNode.get(n);
+        }
+        if (nodeType == ELEMENT_NODE && startsWith(n.tagName, AMP_PREFIX)) {
+          // An AMP node will always have a context node backing it at some
+          // point.
+          return ContextNode.get(n);
+        }
+      }
+      // Navigate up the DOM tree. Notice that we do not automatically go over
+      // a root node boundary.
+      const assignedSlot =
+        /** @type {?Node|undefined} */ (n[ASSIGNED_SLOT_PROP]) ||
+        n.assignedSlot;
+      if (assignedSlot) {
+        n = assignedSlot;
+      } else {
+        n = n.parentNode;
+      }
+    }
+    return null;
```

Comment:
At a glance it appears this will exit with `null` if the given node is in a detached subtree or is itself a root node with `!includeSelf`. Are there other notable cases?

Reply:
Yes: only detached DOM subtree w/o explicit fragment will return `null` here: a document, a shadow root, and a template will all return something.

Reply:
Added an additional comment to clarify this.